### PR TITLE
Include function and parameter in `convert_function` error messages

### PR DESCRIPTION
**Pull request by adpaco-aws**
Include the name of the function and parameter causing `DATA_INVARIANT` to fail in the error messages. This saves time spent in debugging which function and parameter are problematic.

Fixes a couple of mistakes I spotted in the section titles in `COMPILING.md`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/goto_convert_functions.cpp [hunk 1]
```diff
@@ -163,13 +163,36 @@ void goto_convert_functionst::convert_function(
...
+               "parameter identifier must be a known symbol\n"
+               "function: %s\n"
+               "parameter: %s",
+               identifier.c_str(),
+               p.c_str());
+    }
     DATA_INVARIANT(
       symbol_table.has_symbol(p),
-      "parameter identifier must be a known symbol");
+      parameter_error_message);
```
Comment (TGWDB) [inline src/goto-programs/goto_convert_functions.cpp]: I'm not sure that defining and constructing all of this outside the invariant is the right choice. There is also `DATA_INVARIANT_WITH_DIAGNOSTICS` that handles other arguments beyond the message.

**Comment by adpaco-aws** [comment]
Hi @TGWDB , thanks for your suggestion!

I have changed the code to use `DATA_INVARIANT_WITH_DIAGNOSTICS` and print a similar message. Let me know if you see any other issues.

File: src/goto-programs/goto_convert_functions.cpp (hunk 1 above)
Comment (martin-cs) [reply src/goto-programs/goto_convert_functions.cpp]: Yes; what @TGWDB says.

**Review by martin-cs — APPROVED** [review]
Yes; looks good.

Please forgive the pedantry but in future it would be easier if this was two separate PRs.  That way each PR does / fixes one thing.  I realise this is hassle.

**Review by TGWDB — APPROVED** [review]

File: src/goto-programs/goto_convert_functions.cpp [hunk 2]
```diff
@@ -166,10 +166,21 @@ void goto_convert_functionst::convert_function(
...
-    DATA_INVARIANT(!p.empty(), "parameter identifier should not be empty");
-    DATA_INVARIANT(
+    DATA_INVARIANT_WITH_DIAGNOSTICS(
+      !p.empty(),
+      "parameter identifier should not be empty",
+      "function:",
+      identifier,
+      "parameter:",
+      p);
+
```
Comment (kroening) [inline src/goto-programs/goto_convert_functions.cpp:174]: That doesn't make sense: when the message is triggered, `p` is empty. Thus, why log it?

Comment (adpaco-aws) [reply src/goto-programs/goto_convert_functions.cpp:174]: Right. Removed `p` from this case.

**Review by tautschnig — APPROVED** [review]
@adpaco-aws Could you please squash the three commits pertaining to the diagnostic message into a single one?

**Comment by adpaco-aws** [comment]
@tautschnig just squashed them and rebased against develop, let me know if you see any issues.

**Pull request merged**